Gems/AtomTressFX/Code/Passes/HairSkinningComputePass.cpp

```diff
@@ -0,0 +1,277 @@
+                AZStd::lock_guard<AZStd::mutex> lock(m_mutex);
+
+                for (uint32_t j = 0; j < iterations; ++j)
+                {
+                    m_dispatchItems.insert(dispatchItem);
+                }
+            }
+
+            void HairSkinningComputePass::BuildCommandListInternal(const RHI::FrameGraphExecuteContext& context)
+            {
+                if (m_buildShaderAndData)
+                {   // Protect against shader and data async change that were not carried out
+                    m_dispatchItems.clear();
+                    return;
+                }
+
+                RHI::CommandList* commandList = context.GetCommandList();
+
+                // The following will bind all registered Srgs set in m_shaderResourceGroupsToBind
+                // and sends them to the command list ahead of the dispatch.
+                // This includes the PerView, PerScene and PerPass srgs (what about per draw?)
+                SetSrgsForDispatch(commandList);
+
+                for (const RHI::DispatchItem* dispatchItem : m_dispatchItems)
+                {
+                    commandList->Submit(*dispatchItem);
+                }
+
+                // Clear the dispatch items. They will need to be re-populated next frame
+                m_dispatchItems.clear();
+            }
+
+            void HairSkinningComputePass::BuildShaderAndRenderData()
+            {
+                AZStd::lock_guard<AZStd::mutex> lock(m_mutex);
+                if (!AcquireFeatureProcessor())
+                {
+                    m_buildShaderAndData = true;
+                    return;
+                }
+
+                // Flag the FP not to add any more dispatches until shader rebuild was done
+                m_featureProcessor->SetAddDispatchEnable(false);
+
+                while (true)
+                {   // This loop bounds to terminate once the dispatches are submitted.
+                    if (m_dispatchItems.empty())
+                    {
+                        LoadShader();
+
+                        // Flag the FP about the need to rebuild data only after the pass shader was rebuilt
+                        m_featureProcessor->ForceRebuildRenderData();
+                        m_buildShaderAndData = false;
+                        break;
+                    }
+                }
+            }
+
+            // Before reloading shaders, we want to wait for existing dispatches to finish
+            // so shader reloading does not interfere in any way. Because AP reloads are async, there might
+            // be a case where dispatch resources are destructed and will most certainly cause a GPU crash.
+            // If we flag the need for rebuild, the build will be made at the start of the next frame - at
+            // this stage the dispatch items should have been cleared - now we can load the shader and data.
```

Comment:
:) SO the system will actually support that, but I did remove the loop that was attempting to start already this frame and by doing this I'm possibly losing one frame but in the overall scheme it is a cleaner solution.
Since it was handled indirectly in the next frame (via FrameBeginInternal), this is now working with one frame delay.